dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/ComplexBlockSplitting.java

```diff
@@ -41,6 +43,9 @@ private ComplexBlockSplitting() {}
 
     @Override
     public void run(Program program) {
+        // This might not a strong requirement, but if it is weaken, then events created by
+        // this class need to not only get the corresponding oId, but also cId.
+        Preconditions.checkArgument(!program.isCompiled(), "ComplexBlockSplitting should be run before compilation.");
```

Comment:
I think this is a bad idea. We went a long way to get rid of the unnecessary preconditions.
I don't want them back in.
You can copy ALL metadata with `event.copyMetaData` (not sure about the name). This way you don't need to care about when it is actually run. This holds for all the cases you added (I won't mark every single one explicitly).